sfxclient/multitokensink.go

```diff
@@ -0,0 +1,734 @@
+	done         chan bool         // channel to signal that the worker is done
+}
+
+// returns a new instance of worker with an configured emission pipeline
+func newWorker(errorHandler func(error) error, closing chan bool, done chan bool) *worker {
+	w := &worker{
+		lock:         &sync.Mutex{},
+		sink:         NewHTTPSink(),
+		errorHandler: errorHandler,
+		closing:      closing,
+		done:         done,
+	}
+
+	return w
+}
+
+// worker for handling datapoints
+type datapointWorker struct {
+	*worker
+	input     chan *dpMsg // channel for inputing datapoints into a worker
+	buffer    []*datapoint.Datapoint
+	batchSize int
+	stats     *asyncMultiTokenSinkStats // stats about
+}
+
+// emits a series of datapoints
+func (w *datapointWorker) emit(token string) {
+	// set the token on the HTTPSink
+	w.sink.AuthToken = token
+	// emit datapoints and handle any errors
+	err := w.sink.AddDatapoints(context.Background(), w.buffer)
+	w.handleError(err, token, w.buffer)
+	// account for the emitted datapoints
+	w.stats.TotalDatapointsBuffered.Increment(int64(len(w.buffer) * -1))
+}
+
+func (w *datapointWorker) handleError(err error, token string, datapoints []*datapoint.Datapoint) {
+	status := &tokenStatus{
+		status: "unknown",
+		token:  token,
+		val:    int64(len(datapoints)),
+	}
+	if err == nil {
+		status.status = "200"
+	} else {
+		if obj, ok := err.(SFXAPIError); ok {
+			status.status = strconv.Itoa(obj.StatusCode)
+		}
+	}
+	w.stats.TotalDatapointsByToken.Increment(status)
+	if err != nil {
+		_ = w.errorHandler(err)
+	}
+}
+
+func (w *datapointWorker) processMsg(msg *dpMsg) {
+	for len(msg.data) > 0 {
+		msgLength := len(msg.data)
+		remainingBuffer := (w.batchSize - len(w.buffer))
+		if msgLength > remainingBuffer {
+			msgLength = remainingBuffer
+		}
+		w.buffer = append(w.buffer, msg.data[:msgLength]...)
+		msg.data = msg.data[msgLength:]
+		if len(w.buffer) >= w.batchSize {
+			w.emit(msg.token)
+			w.buffer = w.buffer[:0]
+		}
+	}
+}
+
+// bufferDatapoints is responsible for batching incoming datapoints into a buffer
+func (w *datapointWorker) bufferFunc(msg *dpMsg) (stop bool) {
+	var lastTokenSeen = msg.token
+	w.processMsg(msg)
+outer:
+	for len(w.buffer) < w.batchSize {
+		select {
+		case msg = <-w.input:
+			if msg.token != lastTokenSeen {
+				// if the token changes, then emit what ever is in the buffer before proceeding
+				w.emit(lastTokenSeen)
+				w.buffer = w.buffer[:0]
+				lastTokenSeen = msg.token
+			}
+			w.processMsg(msg)
+		default:
+			break outer // emit what ever is in the buffer if there are no more datapoints to read
+		}
+	}
+	// emit the data in the buffer
+	w.emit(msg.token)
+	w.buffer = w.buffer[:0]
+	return
+}
+
+// newBuffer buffers datapoints and events in the pipeline for the duration specified during Startup
+func (w *datapointWorker) newBuffer() {
+	go func() {
+		for {
+			select {
+			case <-w.closing: // check if the worker is in a closing state
```

Comment:
why two channels? if you're doing no work in between just make the w.closing channel have no size (blocking) and when that call returns, you know you've returned.